// @ministryofjustice/hmpps-auth-clients/dist/index.cjs.js
'use strict';

var hmppsRestClient = require('@ministryofjustice/hmpps-rest-client');
var url = require('url');

/**
 * A client for verifying tokens using the HMPPS Token Verification API.
 * @class
 * @extends RestClient
 *
 * @example
 * import VerificationClient from './VerificationClient'
 *
 * const verifyClient = new VerificationClient(
 *   {
 *     enabled: true,
 *     ...ApiConfig
 *   },
 *   console
 * )
 *
 * const isValid = await verifyClient.verifyToken({ user: { username: 'some-user', token: 'abc123' } })
 */
class VerificationClient extends hmppsRestClient.RestClient {
    config;
    logger;
    /**
     * Creates an instance of VerificationClient.
     * @param {VerifyConfig} config - The VerifyConfig settings for the token verification client.
     * @param {Logger|Console} logger - The logging mechanism.
     */
    constructor(config, logger) {
        super('HMPPS Token Verification API', config, logger);
        this.config = config;
        this.logger = logger;
    }
    /**
     * Verifies a token for a given AuthenticatedRequest.
     * @param {AuthenticatedRequest} request - The request object containing user credentials and verification status.
     * @returns {Promise<boolean>} - A promise that resolves to true if valid or if verification is disabled; otherwise false.
     */
    async verifyToken(request) {
        const { user, verified } = request;
        if (!this.config.enabled) {
            this.logger.debug('Token verification disabled, returning token is valid');
            return true;
        }
        if (verified) {
            return true;
        }
        this.logger.debug(`Token request for user "${user.username}"`);
        const response = await this.post({
            path: `/token/verify`,
            data: undefined,
            errorHandler: (path, method, error) => {
                this.logger.error(error, `Error calling tokenVerificationApi ${method} ${path}`);
                return { active: false };
            },
        }, hmppsRestClient.asUser(user.token));
        if (response.active) {
            request.verified = true;
            return true;
        }
        return false;
    }
}

/**
 * An in-memory token store implementation for development or testing.
 * @class
 * @implements {TokenStore}
 *
 * @example
 * import InMemoryTokenStore from './tokenStores/InMemoryTokenStore'
 *
 * const tokenStore = new InMemoryTokenStore()
 * await tokenStore.setToken('some-key', 'abc123', 3600)
 * const token = await tokenStore.getToken('some-key')
 */
class InMemoryTokenStore {
    map = new Map();
    /**
     * Stores a token in memory with a specified expiration.
     * @param {string} key - The key under which to store the token.
     * @param {string} token - The token string.
     * @param {number} durationSeconds - The number of seconds the token remains valid.
     * @returns {Promise<void>}
     */
    async setToken(key, token, durationSeconds) {
        this.map.set(key, { token, expiry: new Date(Date.now() + durationSeconds * 1000) });
        return Promise.resolve();
    }
    /**
     * Retrieves a token from memory by key.
     * @param {string} key - The key associated with the token.
     * @returns {Promise<string | null>} - A promise that resolves with the token or null if not found or expired.
     */
    async getToken(key) {
        if (!this.map.has(key) || this.map.get(key).expiry.getTime() < Date.now()) {
            return Promise.resolve(null);
        }
        return Promise.resolve(this.map.get(key).token);
    }
}

/**
 * A client for authenticating with the HMPPS Auth API, retrieving tokens for system usage or impersonation.
 * @class
 * @extends RestClient
 *
 * @example
 * import AuthenticationClient from './AuthenticationClient'
 *
 * const authClient = new AuthenticationClient(
 *   {
 *     systemClientId: 'client-id',
 *     systemClientSecret: 'client-secret',
 *     ...ApiConfig
 *   },
 *   console
 * )
 *
 * const token = await authClient.getToken('some-user')
 */
class AuthenticationClient extends hmppsRestClient.RestClient {
    config;
    logger;
    tokenStore;
    /**
     * Creates an instance of AuthenticationClient.
     * @param {AuthConfig} config - The AuthConfig settings for the client.
     * @param {Logger|Console} logger - The logging mechanism.
     * @param {TokenStore} [tokenStore=new InMemoryTokenStore()] - Optional token store to cache tokens.
     */
    constructor(config, logger, tokenStore = new InMemoryTokenStore()) {
        super('HMPPS Auth API', config, logger);
        this.config = config;
        this.logger = logger;
        this.tokenStore = tokenStore;
    }
    /**
     * Retrieves a token from the token store or fetches a new one from the HMPPS Auth API.
     * @param {string} [username] - An optional username to impersonate; if omitted, a system (anonymous) token is fetched.
     * @returns {Promise<string>} - A promise that resolves with the retrieved or newly fetched token.
     */
    async getToken(username) {
        const key = username || '%ANONYMOUS%';
        const existingToken = await this.tokenStore.getToken(key);
        if (existingToken) {
            return existingToken;
        }
        const clientToken = Buffer.from(`${this.config.systemClientId}:${this.config.systemClientSecret}`).toString('base64');
        const grantRequest = new url.URLSearchParams({
            grant_type: 'client_credentials',
            ...(username && { username }),
        }).toString();
        this.logger.info(`${grantRequest} HMPPS Auth request for client id '${this.config.systemClientId}'`);
        const response = await this.post({
            path: '/oauth/token',
            headers: {
                Authorization: `Basic ${clientToken}`,
            },
            data: grantRequest,
        });
        const newToken = response.access_token;
        const expiresIn = response.expires_in - 60;
        await this.tokenStore.setToken(key, newToken, expiresIn);
        return newToken;
    }
}

/**
 * A token store implementation using Redis as the backend.
 * The default Redis key prefix is `systemToken`. This can be customised using the optional constructor arg.

 * @class
 * @implements {TokenStore}
 *
 * @example using the default prefix
 * import RedisTokenStore from './tokenStores/RedisTokenStore'
 * import { createClient } from 'redis'
 *
 * const redisClient = createClient() // Use the redis library
 * const tokenStore = new RedisTokenStore(redisClient)
 * await tokenStore.setToken('some-key', 'abc123', 3600)
 * const token = await tokenStore.getToken('some-key')
 *
 * // Redis will have stored the token with the key `systemToken:some-key`
 *
 *
 * @example using a custom prefix
 * import RedisTokenStore from './tokenStores/RedisTokenStore'
 * import { createClient } from 'redis'
 *
 * const redisClient = createClient() // Use the redis library
 * const tokenStore = new RedisTokenStore(redisClient, 'otherTokens')
 * await tokenStore.setToken('some-key', 'abc123', 3600)
 * const token = await tokenStore.getToken('some-key')
 *
 * // Redis will have stored the token with the key `otherTokens:some-key`
 */
class RedisTokenStore {
    client;
    prefix;
    /**
     * Creates an instance of RedisTokenStore.
     * @param {RedisClient} client - A Redis client instance.
     * @param {string} prefix - An optional prefix that will be used for all keys set and retrieved via this token store.
     *                          Default is `systemToken`
     */
    constructor(client, prefix = 'systemToken') {
        this.client = client;
        this.prefix = prefix;
    }
    /**
     * Ensures the Redis client is connected.
     * @private
     * @returns {Promise<void>}
     */
    async ensureConnected() {
        if (!this.client.isOpen) {
            await this.client.connect();
        }
    }
    /**
     * Stores a token in Redis with a specified expiration.
     * @param {string} key - The key under which to store the token.
     * @param {string} token - The token string.
     * @param {number} durationSeconds - The number of seconds the token remains valid.
     * @returns {Promise<void>}
     */
    async setToken(key, token, durationSeconds) {
        await this.ensureConnected();
        await this.client.set(`${this.prefix}:${key}`, token, { EX: durationSeconds });
    }
    /**
     * Retrieves a token from Redis by key.
     * @param {string} key - The key associated with the token.
     * @returns {Promise<string | null>} - A promise that resolves with the token or null if not found or expired.
     */
    async getToken(key) {
        await this.ensureConnected();
        return this.client.get(`${this.prefix}:${key}`);
    }
}

exports.AuthenticationClient = AuthenticationClient;
exports.InMemoryTokenStore = InMemoryTokenStore;
exports.RedisTokenStore = RedisTokenStore;
exports.VerificationClient = VerificationClient;
